### PR TITLE
Implement basic GRANT statement. #1860

### PR DESCRIPTION
**Pull request by mberhault**
Add grammar for simple grant statement. DATABASE level permissions only,
and privileges restricted to ALL (in grammar), or READ|WRITE (in
execution).

There's a decent list of TODOs in both sql/parser/sql.y and
sql/grant.go, some of which are open questions.

The parser part is the one that should be carefully examined.
For example, I know that the grantee_list should not be a list of 'name'
types. There are probably other issues.

Once this is in acceptable shape, I'll do the same for REVOKE, both of
which are needed before we start enforcing permissions.

### CODE REVIEW

File: sql/parser/sql.y [hunk 1]
```diff
@@ -807,6 +814,51 @@ explain_option_arg:
...
+
+privilege_list:
+  name
+  {
+    $$ = []string{$1}
+  }
+  | privilege_list ',' name
+  {
+    $$ = append($1, $3)
+  }
```
Comment (petermattis) [inline sql/parser/sql.y]: This feels way to general. Given that you only support `ALL`, `READ` and `WRITE` at the moment, all of which are already reserved keywords, how about:

```
privilege:
  ALL 
| READ 
| WRITE

privilege_list:
  privilege
| privilegel_list ',' privilege
```

File: sql/grant.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	if n.Target == "" {
+		return nil, errEmptyDatabaseName
+	}
+
+	// Parse privileges.
+	isAll, isRead, isWrite := false, false, false
+	for _, priv := range n.Privileges {
+		switch strings.ToLower(priv) {
+		case "all":
+			isAll = true
```
Comment (petermattis) [inline sql/grant.go]: Can you get rid of `isAll` and just do `isRead, isWrite = true, true`?

File: sql/grant.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		case "read":
+			isRead = true
+		case "write":
+			isWrite = true
+		default:
+			return nil, fmt.Errorf("grant: unknown privilege %q", priv)
+		}
+	}
+
+	if isAll && (isRead || isWrite) {
```
Comment (petermattis) [inline sql/grant.go]: This is already enforced by the grammar and not otherwise tested. I'd just get rid of this, especially if you get rid of `isAll`.

File: sql/grant.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+			writers[grantee] = struct{}{}
+		}
+	}
+
+	// Clear the permissions and re-build from the map.
+	newDesc := *dbDesc
+	newDesc.Read, newDesc.Write = []string{}, []string{}
+	for user := range readers {
+		newDesc.Read = append(newDesc.Read, user)
+	}
```
Comment (petermattis) [inline sql/grant.go]: Might be worthwhile to sort `newDisc.Read` and `newDesc.Write`, otherwise their order will be non-deterministic due to the map iteration.

File: sql/grant.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+	for user := range writers {
+		newDesc.Write = append(newDesc.Write, user)
+	}
+
+	// Now update the descriptor.
+	descKey := keys.MakeDescMetadataKey(newDesc.GetID())
```
Comment (petermattis) [inline sql/grant.go]: `ID` is non-nullable, so `s/GetID()/ID/g`.

File: sql/grant.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	if err := p.db.CPut(descKey, &newDesc, dbDesc); err != nil {
```
Comment (petermattis) [inline sql/grant.go]: Using `CPut` here isn't quite the same as performing everything within a transaction. Someone could come in and change the descriptor and then change it back in between when we read it at the start of this method and when it is written. Not sure if that matters, just pointing it out.

File: sql/grant_test.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	desc := structured.DatabaseDescriptor{}
+	if err := kvDB.GetProto(descKey, &desc); err != nil {
+		t.Fatal(err)
+	}
+	if len(desc.Read) != 1 || !hasString(desc.Read, security.RootUser) {
+		t.Fatalf("wrong Read list: %+v", desc.Read)
+	}
+	if len(desc.Write) != 1 || !hasString(desc.Write, security.RootUser) {
+		t.Fatalf("wrong Write list: %+v", desc.Write)
+	}
```
Comment (petermattis) [inline sql/grant_test.go]: I think the test would be clearer if you added some sort of `SHOW PRIVILEGES` statement. No need to do so in this PR, though.

**Comment by petermattis** [comment]
LGTM

File: sql/parser/grant.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+// Grant represents a GRANT statement.
+type Grant struct {
+	Privileges []string
+	Target     Name
+	Grantees   []string
+}
+
+func (node *Grant) String() string {
+	return fmt.Sprintf("GRANT %s ON %v TO %v", node.Privileges, node.Target, node.Grantees)
```
Comment (tschottdorf) [inline sql/parser/grant.go]: can `Grantees` contain text that messes up the SQL? You might want to escape it. Besides, doesn't seem like `node.Grantees` will format the right way. Shouldn't there be commas involved, etc?

Comment (petermattis) [inline sql/parser/grant.go]: @tschottdorf makes a good point. @MBerhault you need to add appropriate test cases to `ParseTest` to make sure we both parse and output `GRANT` statements correctly.

File: sql/parser/sql.y (hunk 1 above)
Comment (MBerhault) [inline sql/parser/sql.y]: I was following the postgres grammar here, although their reasons may not apply to us:
https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L6211

File: sql/grant.go (hunk 6 above)
Comment (MBerhault) [inline sql/grant.go]: I'm ok with a no-op happening in-between, I just want to error out if things have been changed.

File: sql/grant_test.go (hunk 7 above)
Comment (MBerhault) [inline sql/grant_test.go]: yup. I'll have revoke next, then some way of querying.

File: sql/parser/grant.go (hunk 8 above)
Comment (MBerhault) [inline sql/parser/grant.go]: Will fix the output. How similar do we want it to be? Should it generate exactly the same string?

The grantee is a "name" for now, which is rather strict, and will definitely need to change (see TODO on sql.y, grantee_list). We should figure out what is allowed for usernames. This will be applied in a lot of places.

File: sql/grant.go (hunk 2 above)
Comment (MBerhault) [inline sql/grant.go]: Done.

File: sql/grant.go (hunk 3 above)
Comment (MBerhault) [inline sql/grant.go]: Done.

File: sql/grant.go (hunk 4 above)
Comment (MBerhault) [inline sql/grant.go]: I can, although I'm not sure what it brings us other than making the tests easier, and potentially making lookups faster if we ever have huge permission lists.

File: sql/grant.go (hunk 5 above)
Comment (MBerhault) [inline sql/grant.go]: DOne.

File: sql/grant.go (hunk 4 above)
Comment (MBerhault) [inline sql/grant.go]: actually, done. It'll mean less work on the show, and slightly simpler tests.

File: sql/parser/grant.go (hunk 8 above)
Comment (MBerhault) [inline sql/parser/grant.go]: nevermind, parser_test answers my question.

File: sql/parser/grant.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+type TargetType int
+
+const (
+	TargetDatabase TargetType = iota
+)
+
+var targetNames = map[TargetType]string{
+	TargetDatabase: "DATABASE",
+}
```
Comment (petermattis) [inline sql/parser/grant.go]: Better to use a slice here because `TargetDatabase` is an integer. See `ComparisonOp` and `comparisonOpName` for an example.

Comment (petermattis) [inline sql/parser/grant.go]: You should also add `func (t TargetType) String() string {...}` instead of directly accessing `targetNames` outside of `TargetType` methods.

File: sql/parser/grant.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+// TargetList represents a list of targets of a given type.
+type TargetList struct {
+	Type    TargetType
+	Targets []string
+}
+
+func (node *Grant) String() string {
+	return fmt.Sprintf("GRANT %s ON %s %s TO %v",
+		strings.Join(node.Privileges, ","),
```
Comment (petermattis) [inline sql/parser/grant.go]: The elements of `node.Privileges` are names, but you're not properly quoting them properly. Would be simpler for all of these lists to have the type `NameList` which is a comma separated list of `Names`.

File: sql/parser/sql.y (hunk 1 above)
Comment (petermattis) [inline sql/parser/sql.y]: I don't think it is worth following the postgres grammar precisely here unless we want to follow their privilege model as well (which I don't think we do).

File: sql/grant.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+// Grant gives new privileges to users.
+// Current status:
+// - Target: DATABASE X only
+// - Privileges: ALL or ony of READ, WRITE.
+// TODO(marc): open questions:
+// - should we allow ALL,READ? (ie: should ALL always be by itself?)
+// - should we have root always allowed and not present in the permissions list?
+// - should we make users case-insensitive?
```
Comment (petermattis) [inline sql/grant.go]: I think yes to case-insensitive usernames. We might also want to perform other normalization such as removing `.` (like gmail username normalization).

**Comment by petermattis** [comment]
LGTM

File: sql/parser/grant.go (hunk 9 above)
Comment (MBerhault) [inline sql/parser/grant.go]: added stringer methods for TargetType and TargetList.

File: sql/parser/sql.y (hunk 1 above)
Comment (MBerhault) [inline sql/parser/sql.y]: fair enough. Will fix up.

File: sql/parser/grant.go (hunk 10 above)
Comment (MBerhault) [inline sql/parser/grant.go]: Done.

File: sql/parser/sql.y (hunk 1 above)
Comment (MBerhault) [inline sql/parser/sql.y]: added READ | WRITE.
However, I still need to keep ALL separate. Basically: ALL | "list of READ|WRITE"

File: sql/grant.go (hunk 11 above)
Comment (MBerhault) [inline sql/grant.go]: we'll still need to figure out the regexp for usernames and apply it both here, in the client cert generation, and in the password generation. I'll punt on this for now.

File: sql/parser/grant.go (hunk 10 above)
Comment (MBerhault) [inline sql/parser/grant.go]: hum. with your recent change to encIdent, the names are now quoted, which means that"
```
GRANT READ ON DATABASE foo TO root
```
turns into:
```
GRANT "READ" ON DATABASE foo TO root
```

I could change the statement to build the struct with something other than the keyword strings (eg: list of permission ints), or I could go back to processing the privileges at execution time. The former may be best.

Comment (petermattis) [inline sql/parser/grant.go]: Oh, good point. Yeah, I think it might be best to make the privileges some sort of enumerated type like `TargetType`. But leave that for another PR. This looks good for now.

Comment (petermattis) [inline sql/parser/grant.go]: Heh, just noticed you already added `PrivilegeType`. Nice.

**Comment by petermattis** [comment]
LGTM (again)
